Fix part of #5732: Implement domain logic to show flashback button

subhajitxyz opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
Fixes part of #5732

- Add a new field to the AnswerAndResponse message in exploration.proto.
- Add a logic to detect flashbacks in ExplorationProgressController and add functions in StateDeck.
- Add relevant tests in ExplorationProgressControllerTest.
- Logic gated behind feature flag.

This PR will introduce domain logic for showing new "See example" button (Flashback button).

## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

subhajitxyz committed:
Introduce new field in AnswerAndResponse proto

subhajitxyz committed:
Add logic in ExplorationProgressController and add tests.

subhajitxyz committed:
Fix formatting issue.

subhajitxyz committed:
Update functions and variable names

subhajitxyz committed:
Update name of test function

subhajitxyz committed:
Merge branch 'develop' into flashback-revisit-button-domain-logic

subhajitxyz committed:
Merge branch 'introduce-flashback-feature-flag' into flashback-revisit-button-domain-logic

subhajitxyz committed:
Merge remote-tracking branch 'upstream/develop' into flashback-revisit-button-domain-logic

subhajitxyz committed:
Gated behind feature flag

subhajitxyz committed:
Fix order of imports

subhajitxyz commented:
PTAL @adhiamboperes .

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt:
```diff
@@ -118,7 +120,8 @@ class ExplorationProgressController @Inject constructor(
   @BackgroundDispatcher private val backgroundCoroutineDispatcher: CoroutineDispatcher,
-  private val explorationProgressListeners: Set<@JvmSuppressWildcards ExplorationProgressListener>
+  private val explorationProgressListeners: Set<@JvmSuppressWildcards ExplorationProgressListener>,
+  @EnableFlashbackSupport private val enableFlashbackSupport: PlatformParameterValue<Boolean>,
```
```suggestion
  @EnableFlashbackSupport private val enableFlashbackSupport: PlatformParameterValue<Boolean>
```

subhajitxyz replied:
Done.

adhiamboperes reviewed (changes requested):
Thanks @subhajitxyz!

Please add tests for the feature flag behavior, i.e, expected behaviour when FF is enabled and disabled.

subhajitxyz committed:
Merge remote-tracking branch 'upstream/develop' into flashback-revisit-button-domain-logic

subhajitxyz committed:
fix

subhajitxyz commented:
> Thanks @subhajitxyz!
> 
> Please add tests for the feature flag behavior, i.e, expected behaviour when FF is enabled and disabled.

IIn the ExplorationProgressControllerTest class, all the feature flags are currently enabled inside the setUp() function using the @Before annotation.
To properly test the behavior when feature flag is disabled, we would need to write separate functions for each flag . This would involve setting the desired flag state before each test individually, rather than using the common setUp() method.

A similar approach has already been implemented in StateFragmentTest.

subhajitxyz committed:
Update test exploration json and tests

subhajitxyz committed:
Merge remote-tracking branch 'upstream/develop' into flashback-revisit-button-domain-logic

subhajitxyz committed:
Fix

subhajitxyz committed:
Fix test and add new test in statefragmenttest

subhajitxyz commented:
Hi @adhiamboperes PTAL, I have added new condition for enable flashback button as discussed. Updated test exploration for testing properly.
Update previous tests, and add new one  test in StateFragmentTest.

adhiamboperes reviewed (approved):
This LGTM, thanks @subhajitxyz!
